File: desktop/NumberOne.Core/Services/AuthService.cs
using System.Net;
using System.Net.Http.Json;
using NumberOne.Core.Abstractions;
using NumberOne.Core.Api;
using NumberOne.Core.Models;

namespace NumberOne.Core.Services;

/// <summary>
/// Sign-in, device binding and session restore.
///
/// The binding flow is two-phase, and the reason is worth stating: the server
/// binds inside LoginSerializer.validate(), so a login that carries a device_id
/// binds atomically with no chance to ask first. Binding is a one-way door for
/// the student - only an administrator can undo it - and the design requires a
/// confirmation before it happens.
///
/// device_id is optional on the server, and binding is skipped when it is
/// absent. So the first call omits it and reads StudentProfile.device_id off
/// the response:
///
///   null            -> never bound. Hold the session, show the modal, and on
///                      confirmation log in again *with* the id.
///   ours            -> already bound to this machine. Proceed, no modal.
///   something else  -> bound elsewhere. Show blockedDevice - and, unlike the
///                      failed-login path, we learn which device and when.
///
/// The shared-PC case (this machine already belongs to another student) cannot
/// be detected in advance and surfaces as a 400 on the confirming call.
///
/// Cost is one extra round-trip, on first bind only. The probe issues a token
/// pair that is then discarded; SimpleJWT blacklists refresh tokens on rotation,
/// not on issue, so the unused pair simply expires.
/// </summary>
public sealed class AuthService
{
    private readonly HttpClient _http;
    private readonly ITokenStore _tokens;
    private readonly IDeviceIdentityProvider _device;

    public AuthService(HttpClient http, ITokenStore tokens, IDeviceIdentityProvider device)
    {
        _http = http;
        _tokens = tokens;
        _device = device;
    }

    /// <summary>The signed-in student, once a session exists.</summary>
    public User? CurrentUser { get; private set; }

    /// <summary>
    /// Phase one. Verifies the credentials without binding anything.
    /// </summary>
    public async Task<LoginOutcome> SignInAsync(string username, string password, CancellationToken ct = default)
    {
        TokenPair? probe;
        try
        {
            probe = await PostLoginAsync(
                new LoginRequest { Username = username, Password = password }, ct)
                .ConfigureAwait(false);
        }
        catch (LoginRefusedException ex)
        {
            return new LoginOutcome.Failed(ex.Reason, ex.Message);
        }

        if (probe?.User is null)
            return new LoginOutcome.Failed(LoginFailureReason.Unknown, DesktopMessages.UnexpectedResponse);

        var user = probe.User;

        // This client is the student client. Staff have the web dashboard.
        if (!user.IsStudent)
            return new LoginOutcome.Failed(LoginFailureReason.Unknown, DesktopMessages.StudentsOnly);

        var profile = user.StudentProfile;
        var deviceId = _device.GetDeviceId();

        // Already bound to this machine - the ordinary case after first launch.
        if (profile is not null && string.Equals(profile.DeviceId, deviceId, StringComparison.Ordinal))
        {
            await CommitAsync(probe).ConfigureAwait(false);
            return new LoginOutcome.Success(user);
        }

        // Bound to a different machine.
        if (profile is not null && profile.IsDeviceBound)
        {
            return new LoginOutcome.Failed(
                LoginFailureReason.AccountBoundElsewhere,
                ServerMessages.AccountBoundElsewhere,
                new BoundDeviceInfo(profile.DeviceId, profile.DeviceType, profile.DeviceBoundAt));
        }

        // Never bound. Nothing is persisted until the student confirms.
        return new LoginOutcome.NeedsDeviceBinding(new PendingBind
        {
            Username = username,
            Password = password,
            DeviceId = deviceId,
            DeviceType = _device.GetDeviceType(),
            MachineName = _device.GetMachineName(),
            User = user,
        });
    }

    /// <summary>
    /// Phase two. Logs in again carrying the device id, which is what actually
    /// binds. Call only after the student confirmed the modal.
    /// </summary>
    public async Task<LoginOutcome> ConfirmBindAsync(PendingBind pending, CancellationToken ct = default)
    {
        TokenPair? bound;
        try
        {
            bound = await PostLoginAsync(new LoginRequest
            {
                Username = pending.Username,
                Password = pending.Password,
                DeviceId = pending.DeviceId,
                DeviceType = pending.DeviceType,
            }, ct).ConfigureAwait(false);
        }
        catch (LoginRefusedException ex)
        {
            // The expected refusal here is DeviceBoundToAnotherStudent: between
            // the probe and now we asked the server to bind, and it found this
            // machine already held by someone else. Common on a family or lab PC.
            return new LoginOutcome.Failed(ex.Reason, ex.Message);
        }

        if (bound?.User is null)
            return new LoginOutcome.Failed(LoginFailureReason.Unknown, DesktopMessages.UnexpectedResponse);

        await CommitAsync(bound).ConfigureAwait(false);
        return new LoginOutcome.Success(bound.User);
    }

    /// <summary>
    /// Restores a session at launch from stored tokens, so a student who never
    /// signed out lands in the app rather than at a login form.
    ///
    /// A 401 here is already handled by AuthenticatingHandler, which will have
    /// tried a refresh before this sees the failure — so a rejection at this
    /// point means the refresh token is dead too, not merely that the access
    /// token expired overnight.
    ///
    /// The four outcomes are deliberately distinct. Collapsing "offline" into
    /// "rejected" is the trap: it sends a student with perfectly good tokens to
    /// a login form that cannot reach the server either, which is a dead end
    /// they can only escape by finding a network.
    /// </summary>
    public async Task<SessionRestore> RestoreSessionAsync(CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(await _tokens.GetAccessTokenAsync().ConfigureAwait(false)))
            return SessionRestore.None;

        try
        {
            using var response = await _http.GetAsync(ApiEndpoints.Me, ct).ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
                return await RejectAsync().ConfigureAwait(false);

            var user = await response.Content
                .ReadFromJsonAsync<User>(ApiClientFactory.Json, ct)
                .ConfigureAwait(false);

            if (user is null || !user.IsStudent)
                return await RejectAsync().ConfigureAwait(false);

            // The binding is checked on every launch, not only at sign-in.
            //
            // An administrator can unbind an account, and it can then be bound
            // to another machine. The tokens on THIS machine stay valid through
            // all of that, so without this check a student who had been moved
            // to a new computer would keep a working session on the old one —
            // which is the whole thing the one-device rule exists to prevent.
            var bound = user.StudentProfile?.DeviceId;
            var here = SafeDeviceId();

            // Only compared when BOTH are known. A machine whose hardware id
            // cannot be read must not be locked out of its own session over a
            // comparison that could not be made — the sign-in path already
            // refuses to bind without an id, so nothing is weakened by being
            // permissive here.
            if (!string.IsNullOrWhiteSpace(bound) &&
                !string.IsNullOrWhiteSpace(here) &&
                !string.Equals(bound, here, StringComparison.Ordinal))
            {
                return await RejectAsync().ConfigureAwait(false);
            }

            CurrentUser = user;
            return SessionRestore.Restored;
        }
        catch (Exception ex) when (ApiAttempt.IsTransport(ex))
        {
            // Offline at launch. The tokens are probably fine, so they stay:
            // the caller opens the app and every section shows its own retry.
            // If they turn out to be dead, the first request that gets through
            // raises SessionExpired and the student is sent to login then.
            return SessionRestore.Unverified;
        }
    }

    /// <summary>
    /// The tokens are no good. Clearing them is the point — leaving a dead pair
    /// on a shared machine is exactly what sign-out exists to avoid.
    /// </summary>
    private async Task<SessionRestore> RejectAsync()
    {
        CurrentUser = null;
        await _tokens.ClearAsync().ConfigureAwait(false);

        return SessionRestore.Rejected;
    }

    /// <summary>
    /// The hardware id, or null. A provider that cannot read the machine throws;
    /// that must not stop a launch, and a null simply skips the binding check
    /// rather than locking the student out of their own app.
    /// </summary>
    private string? SafeDeviceId()
    {
        try { return _device.GetDeviceId(); }
        catch (Exception) { return null; }
    }

    /// <summary>
    /// Signs out. Blacklists the refresh token server-side when reachable, and
    /// clears local storage either way - a failed logout call must never leave
    /// tokens on the machine.
    /// </summary>
    public async Task SignOutAsync(CancellationToken ct = default)
    {
        var refresh = await _tokens.GetRefreshTokenAsync().ConfigureAwait(false);

        if (!string.IsNullOrWhiteSpace(refresh))
        {
            try
            {
                using var _ = await _http.PostAsJsonAsync(
                    ApiEndpoints.Logout, new RefreshRequest { Refresh = refresh }, ApiClientFactory.Json, ct)
                    .ConfigureAwait(false);
            }
            catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
            {
                // Offline. Clearing locally is the part that matters.
            }
        }

        CurrentUser = null;
        await _tokens.ClearAsync().ConfigureAwait(false);
    }

    private async Task<TokenPair?> PostLoginAsync(LoginRequest request, CancellationToken ct)
    {
        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsJsonAsync(ApiEndpoints.Login, request, ApiClientFactory.Json, ct)
                .ConfigureAwait(false);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new LoginRefusedException(LoginFailureReason.Network, DesktopMessages.NoConnection);
        }

        using (response)
        {
            if (response.IsSuccessStatusCode)
                return await response.Content
                    .ReadFromJsonAsync<TokenPair>(ApiClientFactory.Json, ct)
                    .ConfigureAwait(false);

            var body = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
            var message = DrfError.ExtractMessage(body);

            // 400 is the only refusal LoginSerializer produces; anything else is
            // a server or proxy fault and should not be dressed up as a
            // credential problem.
            if (response.StatusCode != HttpStatusCode.BadRequest)
                throw new LoginRefusedException(
                    LoginFailureReason.Unknown,
                    message ?? DesktopMessages.ServerFault);

            var reason = LoginFailureClassifier.Classify(message);
            throw new LoginRefusedException(reason, message ?? DesktopMessages.UnexpectedResponse);
        }
    }

    private async Task CommitAsync(TokenPair pair)
    {
        await _tokens.SaveAsync(pair.Access, pair.Refresh).ConfigureAwait(false);
        CurrentUser = pair.User;
    }

    private sealed class LoginRefusedException : Exception
    {
        public LoginFailureReason Reason { get; }

        public LoginRefusedException(LoginFailureReason reason, string message) : base(message)
            => Reason = reason;
    }
}

/// <summary>
/// Strings the client owns, for conditions the server has no message for.
/// Everything the server does have a message for is shown verbatim.
/// </summary>
public static class DesktopMessages
{
    public const string NoConnection =
        "لا يوجد اتصال بالإنترنت. تحقّق من الشبكة ثم أعد المحاولة.";

    public const string ServerFault =
        "تعذّر الاتصال بالخادم. يرجى المحاولة لاحقاً.";

    public const string UnexpectedResponse =
        "استجابة غير متوقعة من الخادم. يرجى المحاولة لاحقاً.";

    /// <summary>
    /// A section failed for a reason that is not the network.
    ///
    /// Worded so it can never be mistaken for being offline: a student who
    /// retries a connection problem is doing the right thing, and a student who
    /// retries a defect is not.
    /// </summary>
    public const string SectionFailed =
        "تعذّر عرض هذا القسم. إن تكرر الأمر فأبلغ إدارة المدرسة.";

    public const string StudentsOnly =
        "هذا التطبيق مخصص للطلاب فقط. يرجى استخدام لوحة التحكم عبر المتصفح.";

    public const string DeviceIdUnavailable =
        "تعذّر تحديد معرّف هذا الجهاز. يرجى التواصل مع الإدارة.";
}

File: desktop-tauri/src/auth/authService.ts
import { invoke } from "@tauri-apps/api/core";

import { ApiError, Messages, NetworkError, request } from "../api/client";
import { Endpoints } from "../api/endpoints";
import { classifyLoginFailure, LoginFailureReason, ServerMessages } from "./messages";
import { tokens } from "./tokens";

/**
 * Sign-in, device binding and session restore.
 *
 * THE BINDING FLOW IS TWO-PHASE, and the reason is worth stating: the server
 * binds inside LoginSerializer.validate(), so a login carrying a device_id
 * binds atomically with no chance to ask first. Binding is a one-way door for
 * the student — only an administrator can undo it — and the design requires
 * confirmation before it happens.
 *
 * device_id is optional server-side and binding is skipped when it is absent.
 * So the first call omits it and reads StudentProfile.device_id off the reply:
 *
 *   null           → never bound. Hold the session, show the dialog, and on
 *                    confirmation log in again WITH the id.
 *   ours           → already bound to this machine. Proceed, no dialog.
 *   something else → bound elsewhere. Show the blocked screen — and unlike the
 *                    failed-login path, we learn which device and when.
 *
 * The shared-PC case (this machine already belongs to another student) cannot
 * be detected in advance and surfaces as a 400 on the confirming call.
 *
 * Cost is one extra round trip, on first bind only. The probe issues a token
 * pair that is then discarded; SimpleJWT blacklists refresh tokens on rotation
 * rather than on issue, so the unused pair simply expires.
 */

export interface StudentProfile {
  device_id?: string | null;
  device_type?: string | null;
  device_bound_at?: string | null;
  is_device_bound?: boolean;
  [key: string]: unknown;
}

export interface User {
  id: number;
  username: string;
  role?: string | null;
  student_profile?: StudentProfile | null;
  [key: string]: unknown;
}

export interface DeviceIdentity {
  id: string;
  device_type: string;
  machine_name: string;
}

export interface BoundDeviceInfo {
  deviceId: string | null;
  deviceType: string | null;
  boundAt: string | null;
}

/** A verified sign-in held back until the student confirms the bind. */
export interface PendingBind {
  username: string;
  /** Carried because binding needs a second login call; the server binds during login. */
  password: string;
  deviceId: string;
  deviceType: string;
  machineName: string;
  user: User;
}

export type LoginOutcome =
  | { kind: "success"; user: User }
  | { kind: "needs-device-binding"; pending: PendingBind }
  | { kind: "failed"; reason: LoginFailureReason; message: string; boundDevice?: BoundDeviceInfo };

/**
 * What a launch found in the token store.
 *
 * Four cases rather than a boolean, because two of them look like failure and
 * must not be treated alike. "rejected" means the session is over; "unverified"
 * means the app could not ask, which is not the student's problem and must not
 * cost them their session.
 */
export type SessionRestore = "none" | "restored" | "rejected" | "unverified";

function isStudent(user: User): boolean {
  return user.role === "student" || user.student_profile != null;
}

async function deviceIdentity(): Promise<DeviceIdentity | null> {
  try {
    return await invoke<DeviceIdentity>("device_identity");
  } catch {
    // A machine whose hardware id cannot be read must not be locked out of its
    // own session over a comparison that could not be made. The sign-in path
    // refuses to bind without an id, so nothing is weakened by being permissive.
    return null;
  }
}

class AuthService {
  private user: User | null = null;

  get currentUser(): User | null {
    return this.user;
  }

  /** Phase one. Verifies the credentials without binding anything. */
  async signIn(username: string, password: string): Promise<LoginOutcome> {
    let probe: { access: string; refresh: string; user?: User };

    try {
      probe = await request(Endpoints.login, { body: { username, password } });
    } catch (error) {
      return this.refusal(error);
    }

    const user = probe.user;
    if (!user) {
      return { kind: "failed", reason: "unknown", message: Messages.unexpectedResponse };
    }

    // This client is the student client. Staff have the web dashboard.
    if (!isStudent(user)) {
      return { kind: "failed", reason: "unknown", message: Messages.studentsOnly };
    }

    const device = await deviceIdentity();
    if (!device) {
      return { kind: "failed", reason: "unknown", message: Messages.deviceIdUnavailable };
    }

    const profile = user.student_profile ?? null;

    // Already bound to this machine — the ordinary case after first launch.
    if (profile && profile.device_id === device.id) {
      await this.commit(probe.access, probe.refresh, user);
      return { kind: "success", user };
    }

    // Bound to a different machine.
    if (profile && profile.is_device_bound) {
      return {
        kind: "failed",
        reason: "account-bound-elsewhere",
        message: ServerMessages.accountBoundElsewhere,
        boundDevice: {
          deviceId: profile.device_id ?? null,
          deviceType: profile.device_type ?? null,
          boundAt: profile.device_bound_at ?? null,
        },
      };
    }

    // Never bound. Nothing is persisted until the student confirms.
    return {
      kind: "needs-device-binding",
      pending: {
        username,
        password,
        deviceId: device.id,
        deviceType: device.device_type,
        machineName: device.machine_name,
        user,
      },
    };
  }

  /**
   * Phase two. Logs in again carrying the device id, which is what binds.
   * Call only after the student confirmed.
   */
  async confirmBind(pending: PendingBind): Promise<LoginOutcome> {
    let bound: { access: string; refresh: string; user?: User };

    try {
      bound = await request(Endpoints.login, {
        body: {
          username: pending.username,
          password: pending.password,
          device_id: pending.deviceId,
          device_type: pending.deviceType,
        },
      });
    } catch (error) {
      // The expected refusal here is device-bound-to-another-student: between
      // the probe and now we asked the server to bind, and it found this
      // machine already held by someone else. Common on a family or lab PC.
      return this.refusal(error);
    }

    if (!bound.user) {
      return { kind: "failed", reason: "unknown", message: Messages.unexpectedResponse };
    }

    await this.commit(bound.access, bound.refresh, bound.user);
    return { kind: "success", user: bound.user };
  }

  /**
   * Restores a session at launch, so a student who never signed out lands in
   * the app rather than at a login form.
   *
   * A 401 here has already been through one refresh in the client, so a
   * rejection at this point means the refresh token is dead too — not merely
   * that the access token expired overnight.
   */
  async restoreSession(): Promise<SessionRestore> {
    const stored = await tokens.get();
    if (!stored.access) return "none";

    let user: User;
    try {
      user = await request<User>(Endpoints.me);
    } catch (error) {
      if (error instanceof NetworkError) {
        // Offline at launch. The tokens are probably fine, so they stay: the
        // caller opens the app and every section shows its own retry. If they
        // do turn out to be dead, the first request that gets through raises
        // session-expired and the student goes to login then.
        return "unverified";
      }
      return this.reject();
    }

    if (!isStudent(user)) return this.reject();

    // The binding is checked on every launch, not only at sign-in.
    //
    // An administrator can unbind an account, and it can then be bound to
    // another machine. The tokens on THIS machine stay valid through all of
    // that — so without this check a student who had been moved to a new
    // computer would keep a working session on the old one, which is the whole
    // thing the one-device rule exists to prevent.
    const bound = user.student_profile?.device_id;
    const here = (await deviceIdentity())?.id;

    // Compared only when BOTH are known, for the reason in deviceIdentity().
    if (bound && here && bound !== here) return this.reject();

    this.user = user;
    return "restored";
  }

  /**
   * Blacklists the refresh token server-side when reachable, and clears local
   * storage either way — a failed logout call must never leave tokens behind.
   */
  async signOut(): Promise<void> {
    const refresh = (await tokens.get()).refresh;

    if (refresh) {
      try {
        await request(Endpoints.logout, { body: { refresh } });
      } catch {
        // Offline, or already blacklisted. Clearing locally is the part that
        // matters on a shared machine.
      }
    }

    this.user = null;
    await tokens.clear();
  }

  private async commit(access: string, refresh: string, user: User): Promise<void> {
    await tokens.save(access, refresh);
    this.user = user;
  }

  private async reject(): Promise<SessionRestore> {
    this.user = null;
    await tokens.clear();
    return "rejected";
  }

  private refusal(error: unknown): LoginOutcome {
    if (error instanceof NetworkError) {
      return { kind: "failed", reason: "network", message: Messages.noConnection };
    }

    if (error instanceof ApiError) {
      // 400 is the only refusal LoginSerializer produces. Anything else is a
      // server or proxy fault and should not be dressed up as a credential
      // problem.
      if (error.status !== 400) {
        return { kind: "failed", reason: "unknown", message: error.message };
      }
      return { kind: "failed", reason: classifyLoginFailure(error.message), message: error.message };
    }

    return { kind: "failed", reason: "unknown", message: Messages.unexpectedResponse };
  }
}

export const auth = new AuthService();
